File: docs/source/src/python/fitness.py
# --8<-- [start:rastrigin]
import radiate as rd
import math

A = 10.0
RANGE = 5.12
N_GENES = 2


def fit(x: list[float]) -> float:
    value = A * N_GENES
    for i in range(N_GENES):
        value += x[i] ** 2 - A * math.cos((2.0 * 3.141592653589793 * x[i]))
    return value


codec = rd.FloatCodec(N_GENES, init_range=(-RANGE, RANGE))
engine = rd.Engine(codec).fitness(fit).minimizing()
# --8<-- [end:rastrigin]

# --8<-- [start:fitness_decorator]
import radiate as rd


@rd.fitness  # <- this decorator
def fit_decorated(x: list[float]) -> float:
    value = A * N_GENES
    for i in range(N_GENES):
        value += x[i] ** 2 - A * math.cos((2.0 * 3.141592653589793 * x[i]))
    return value


# ... fill in the rest of your engine setup as normal (check first example above)...
# --8<-- [end:fitness_decorator]

# --8<-- [start:batch_fitness]
import radiate as rd
import math

A = 10.0
RANGE = 5.12
N_GENES = 2


# NOTE this function expects a batch of inputs and returns a batch of outputs
# the order in which the inputs are given is the order in which the outputs are returned
def fit_batch(x: list[list[float]]) -> list[float]:
    assert len(x) > 1

    results = []
    for member in x:
        value = A * N_GENES
        for i in range(N_GENES):
            value += member[i] ** 2 - A * math.cos(
                (2.0 * 3.141592653589793 * member[i])
            )
        results.append(value)
    return results


codec = rd.FloatCodec(N_GENES, init_range=(-RANGE, RANGE))

# Create the genetic engine with batch fitness function.
# Just wrap your fitness function in 'rd.BatchFitness'
engine = rd.Engine(codec).fitness(rd.BatchFitness(fit_batch)).minimizing()
# --8<-- [end:batch_fitness]

# --8<-- [start:batch_fitness_decorator]
import radiate as rd
import math

A = 10.0
RANGE = 5.12
N_GENES = 2


# NOTE this function expects a batch of inputs and returns a batch of outputs
# the order in which the inputs are given is the order in which the outputs are returned
@rd.fitness(batch=True)  # <- this decorator with batch=True
def fit_batch_decorated(x: list[list[float]]) -> list[float]:
    assert len(x) > 1

    results = []
    for member in x:
        value = A * N_GENES
        for i in range(N_GENES):
            value += member[i] ** 2 - A * math.cos(
                (2.0 * 3.141592653589793 * member[i])
            )
        results.append(value)
    return results


codec = rd.FloatCodec(N_GENES, init_range=(-RANGE, RANGE))

# Create the genetic engine with batch fitness function.
# NOTE: We no longer need to wrap 'fitness_fn' in 'rd.BatchFitness'
engine = rd.Engine(codec).fitness(fit_batch_decorated).minimizing()
# --8<-- [end:batch_fitness_decorator]

# --8<-- [start:novelty_search]
import radiate as rd


# Define a behavioral descriptor
def behavior(individual: list[float]) -> list[float]:
    # Return behavioral characteristics
    # Some code that describes the behavior of a vector
    # The individual here is a list[float] because we are using a FloatCodec vector below
    ...


# Create novelty search fitness function
novelty_fitness = rd.NoveltySearch(
    descriptor=behavior,
    # can use any of the distance inputs. The engine will use this to
    # determine how 'novel' an individual is compared to the other's in the
    # archinve or population, ultimently resulting in the individuals fitness score.
    distance=rd.CosineDistance(),  # Distance metric to use,
    k=10,  # Number of nearest neighbors to consider
    threshold=0.1,  # Novelty threshold for archive addition
    archive_size=1000,  # defaults to 1000
)

engine = (
    # The decoded value of your codec (a list[float] in this case) will be fed into the `behavior` function of your NoveltySearch fitness_func
    rd.Engine.float(10, init_range=(0, 10), bounds=(0, 10))
    .fitness(novelty_fitness)
    .maximizing()  # We want to maximize novelty - however this is the default so its not necessary to define
)
# --8<-- [end:novelty_search]

# --8<-- [start:novelty_decorator]
import radiate as rd


# Define a behavioral descriptor
@rd.novelty(distance=rd.Dist.cosine(), k=10, threshold=0.1, archive=1000)
def behavior_decorated(individual: list[float]) -> list[float]:
    # Return behavioral characteristics
    # Some code that describes the behavior of a vector
    # The individual here is a list[float] because we are using a FloatCodec vector below
    ...


engine = rd.Engine.float(10, init_range=(0, 10)).fitness(behavior_decorated)
# --8<-- [end:novelty_decorator]

File: source/src/rust/fitness.rs
use radiate::prelude::*;

// Stand-ins for the "plug in your own model" sections (Composite & Novelty).
// In a real program these would be your decoded domain type and its codec.
type MyDecodedModel = Vec<f32>;

#[derive(Clone)]
struct MyModel(Vec<f32>);

impl MyModel {
    fn get_behavior_vector(&self) -> Vec<f32> {
        self.0.clone()
    }
}

fn main() {
    // --8<-- [start:rastrigin]
    const N_GENES: usize = 2;
    const A: f32 = 10.0;

    fn rastrigin_function(genotype: Vec<f32>) -> f32 {
        let mut value = A * N_GENES as f32;
        for i in 0..N_GENES {
            value += genotype[i].powi(2) - A * (2.0 * std::f32::consts::PI * genotype[i]).cos();
        }

        value
    }

    let engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(N_GENES, -5.12..5.12))
        .minimizing()
        .fitness_fn(rastrigin_function)
        .build();
    // --8<-- [end:rastrigin]

    // --8<-- [start:batch]
    let engine = GeneticEngine::builder()
        .codec(IntChromosome::from((5, 0..100)))
        // Replace the original 'fitness_fn' call with 'batch_fitness_fn' to enable batch fitness.
        // This fitness function will receive a single batch containing all individuals which need evaluation
        .batch_fitness_fn(|phenotypes: &[Vec<i32>]| {
            phenotypes
                .iter()
                .map(|geno| geno.iter().sum::<i32>())
                .collect()
        })
        .build();

    // When using an parallel executor, the batches will be grouped together to evaluate on separate threads
    let engine = GeneticEngine::builder()
        .codec(IntChromosome::from((5, 0..100)))
        // 'batch_fitness_fn' will receive 7 batches of individuals during each generation's evaluation
        .executor(Executor::FixedSizedWorkerPool(7))
        .batch_fitness_fn(|phenotypes: &[Vec<i32>]| {
            phenotypes
                .iter()
                .map(|geno| geno.iter().sum::<i32>())
                .collect()
        })
        .build();
    // --8<-- [end:batch]

    // --8<-- [start:raw]
    fn my_fitness_fn(genotype: &Genotype<FloatChromosome<f32>>) -> f32 {
        // Evaluate the genotype directly without decoding
        genotype
            .iter()
            .map(|chromosome| chromosome.iter().map(|gene| *gene.allele()).sum::<f32>())
            .sum::<f32>()
    }

    let mut engine = GeneticEngine::builder()
        .codec(FloatCodec::vector(10, 0.0..1.0))
        .raw_fitness_fn(my_fitness_fn)
        // or .raw_batch_fitness_fn(...) for batch raw fitness functions
        // ... other parameters ...
        .build();

    // Run the engine
    let result = engine.run(|epoch| epoch.index() >= 100);
    // --8<-- [end:raw]

    let my_model_codec = FloatCodec::vector(10, 0.0..1.0); // decodes to MyDecodedModel (Vec<f32>)

    // --8<-- [start:composite]
    fn accuracy_objective(model: &MyDecodedModel) -> f32 {
        // ... calculate accuracy ...
        model.iter().sum()
    }

    fn complexity_objective(model: &MyDecodedModel) -> f32 {
        // ... calculate complexity ...
        model.len() as f32
    }

    fn efficiency_objective(model: &MyDecodedModel) -> f32 {
        // ... calculate efficiency ...
        model.iter().copied().fold(0.0, f32::max)
    }

    // Create weighted composite fitness function - this version computes a
    // weighted average of each function given to it
    let composite_fitness = CompositeFitnessFn::new()
        .add_weighted_fn(accuracy_objective, 0.6) // 60% weight on accuracy
        .add_weighted_fn(complexity_objective, 0.25) // 25% weight on complexity
        .add_weighted_fn(efficiency_objective, 0.15); // 15% weight on efficiency

    // Create an equal weight composite fitness function with equal weights.
    // Meaning these are all essentially just added together to create a single objective.
    let composite_fitness = CompositeFitnessFn::new()
        .add_fitness_fn(accuracy_objective)
        .add_fitness_fn(complexity_objective)
        .add_fitness_fn(efficiency_objective);

    // Add it to the engine like any other fitness function
    let engine = GeneticEngine::builder()
        .codec(my_model_codec)
        .fitness_fn(composite_fitness)
        .build();
    // --8<-- [end:composite]

    // A codec whose decoded type is `MyModel` so the novelty descriptor below can read it.
    let my_model_codec = FnCodec::new()
        .with_encoder(|| Genotype::from(FloatChromosome::from((10, 0.0..1.0))))
        .with_decoder(|geno: &Genotype<FloatChromosome<f32>>| {
            MyModel(geno[0].iter().map(|gene| *gene.allele()).collect())
        });

    // --8<-- [start:novelty]
    // Define a behavioral descriptor
    struct MyModelBehaviorDescriptor;

    // ... rest of impl ...

    impl Novelty<MyModel> for MyModelBehaviorDescriptor {
        fn description(&self, individual: &MyModel) -> Vec<f32> {
            // Return behavioral characteristics (e.g., outputs on test cases)
            individual.get_behavior_vector()
        }
    }

    // Create novelty search fitness function
    let novelty_fitness = NoveltySearch::new(MyModelBehaviorDescriptor)
        .k(10)
        .threshold(0.1)
        .archive_size(1000) // Optional: set archive size - default is 1000
        .cosine_distance(); // Optional set the distance parameter used
    // .euclidean_distance() // euclidean_distance is the default
    // .hamming_distance()

    // Novelty is also implemented for any F where F: Fn(&T) -> Vec<f32>. Meaning, you can
    // just as easily feed a function to NoveltySearch as long as it takes a borrowed T (&T)
    // and returns a Vec<f32>
    let function_novelty_fitness =
        NoveltySearch::new(|individual: &MyModel| individual.get_behavior_vector());

    let engine = GeneticEngine::builder()
        // The decoded genotype from your codec (my_model_codec in this case) will be fed
        // into the `description` function from the Novelty trait impl
        .codec(my_model_codec)
        .maximizing()
        .fitness_fn(novelty_fitness)
        .build();
    // --8<-- [end:novelty]
}
